File: nyaya/packages/shared/decision_contract.py
from pydantic import BaseModel, Field, validator
from typing import Dict, Any, List
from enum import Enum

class EnforcementState(str, Enum):
    CLEAR = "clear"
    BLOCK = "block"
    ESCALATE = "escalate"
    SOFT_REDIRECT = "soft_redirect"
    CONDITIONAL = "conditional"

class EnforcementVerdict(str, Enum):
    ENFORCEABLE = "ENFORCEABLE"
    PENDING_REVIEW = "PENDING_REVIEW"
    NON_ENFORCEABLE = "NON_ENFORCEABLE"

class EnforcementStatus(BaseModel):
    state: EnforcementState
    verdict: EnforcementVerdict
    reason: str = ""
    barriers: List[str] = []
    blocked_path: str = None
    escalation_required: bool = False
    escalation_target: str = None
    redirect_suggestion: str = None
    safe_explanation: str = ""
    trace_id: str

    @validator('trace_id')
    def trace_id_must_be_present(cls, v):
        if not v or not isinstance(v, str):
            raise ValueError('trace_id must be a non-empty string')
        return v

class DecisionContract(BaseModel):
    """
    Canonical Decision Contract - Single source of truth for Nyaya ecosystem.

    This immutable schema MUST be used across:
    - Backend response models
    - Frontend state management
    - Observer pipeline JSON output
    - All data exchange points
    """
    trace_id: str = Field(..., description="UUID trace identifier")
    jurisdiction: str = Field(..., description="Legal jurisdiction code")
    domain: str = Field(..., description="Legal domain")
    legal_route: List[str] = Field(..., description="Sequence of agents/legal steps")
    reasoning_trace: Dict[str, Any] = Field(..., description="Detailed reasoning process")
    enforcement_status: EnforcementStatus = Field(..., description="Enforcement state and verdict")
    confidence: float = Field(..., ge=0.0, le=1.0, description="Confidence score")

    @validator('trace_id', 'jurisdiction', 'domain')
    def non_empty_strings(cls, v):
        if not v or not isinstance(v, str) or len(v.strip()) == 0:
            raise ValueError('Field must be a non-empty string')
        return v.strip()

    @validator('legal_route')
    def legal_route_non_empty(cls, v):
        if not v or len(v) == 0:
            raise ValueError('legal_route must be a non-empty list')
        return v

    class Config:
        extra = 'forbid'  # No extra fields allowed

# Validation function for programmatic gatekeeper
def validate_decision_contract(data: Dict[str, Any]) -> DecisionContract:
    """
    Validates data against DecisionContract schema.
    Raises ValidationError if invalid.
    """
    return DecisionContract(**data)

File: nyaya/packages/shared/decision_contract.ts
import { z } from 'zod'

// Zod schemas for Decision Contract validation

export const EnforcementStateSchema = z.enum([
  'clear',
  'block',
  'escalate',
  'soft_redirect',
  'conditional'
])

export const EnforcementVerdictSchema = z.enum([
  'ENFORCEABLE',
  'PENDING_REVIEW',
  'NON_ENFORCEABLE'
])

export const EnforcementStatusSchema = z.object({
  state: EnforcementStateSchema,
  verdict: EnforcementVerdictSchema,
  reason: z.string().default(''),
  barriers: z.array(z.string()).default([]),
  blocked_path: z.string().nullable().optional(),
  escalation_required: z.boolean().default(false),
  escalation_target: z.string().nullable().optional(),
  redirect_suggestion: z.string().nullable().optional(),
  safe_explanation: z.string().default(''),
  trace_id: z.string().min(1, 'trace_id must be non-empty')
})

export const DecisionContractSchema = z.object({
  trace_id: z.string().min(1, 'trace_id must be non-empty'),
  jurisdiction: z.string().min(1, 'jurisdiction must be non-empty'),
  domain: z.string().min(1, 'domain must be non-empty'),
  legal_route: z.array(z.string()).min(1, 'legal_route must be non-empty'),
  reasoning_trace: z.record(z.any()),
  enforcement_status: EnforcementStatusSchema,
  confidence: z.number().min(0).max(1)
}).strict() // No extra fields allowed

// TypeScript interfaces
export interface EnforcementStatus {
  state: 'clear' | 'block' | 'escalate' | 'soft_redirect' | 'conditional'
  verdict: 'ENFORCEABLE' | 'PENDING_REVIEW' | 'NON_ENFORCEABLE'
  reason?: string
  barriers?: string[]
  blocked_path?: string | null
  escalation_required?: boolean
  escalation_target?: string | null
  redirect_suggestion?: string | null
  safe_explanation?: string
  trace_id: string
}

export interface DecisionContract {
  trace_id: string
  jurisdiction: string
  domain: string
  legal_route: string[]
  reasoning_trace: Record<string, any>
  enforcement_status: EnforcementStatus
  confidence: number
}

// Validation function for programmatic gatekeeper
export function validateDecisionContract(data: unknown): DecisionContract {
  return DecisionContractSchema.parse(data)
}

// Type guard
export function isDecisionContract(data: unknown): data is DecisionContract {
  return DecisionContractSchema.safeParse(data).success
}